book/src/exec.md

```diff
@@ -0,0 +1,168 @@
+
+**Step 3**: A solver sees Alice's ptx and Bob's ptx, matches them together, and creates a new partial transaction. Alice's intentVP is satisfied, and her intent note is spent. 
+
+Total per-token balances:
+
+|token|spent|output|spent - output|
+|-|-|-|-|
+|star NFT|1|0|1|
+|blue dolphin NFT|1|1||
+|blue intent token|[1] - [1]|-|0|
+|yellow intent token|[1] |-|[1]|
+
+**Step 4**: A solver sees all previous partial transactions and the initial transaction created by Charlie. The solver matches them together and creates new partial transactions, sending the tree to Bob and the star to Charlie. VPs of Bob and Charlie are now satisfied, the corresponding intent notes are spent. The  per-token balance of partial transactions is equal to zero, which means it is possible to create a transaction.
+Total per-token balances:
+
+|token|spent|output|spent - output|
+|-|-|-|-|
+|star NFT|1|1|0|
+|blue dolphin NFT|1|1|0|
+|blue intent token|[1] - [1] = [0]|-|0|
+|yellow intent token|[1] - [1]|-|0|
+|tree NFT|1|1|0|
+|green intent token|[1] - [1]|-|0|
+
+**Step 5**:
+The final transaction containing the spent and output notes from partial transactions is created with all proofs attached.
+
+### 2. Complex intentVP
+
+Let's consider the situation where one of the parties has a complex VP (compared to the VPs above). Here Alice has two notes: [1] of token A and [2] of token B and wants to get a blue dolphin NFT in  exchange for one of them and get the other one back. Bob has a simple intentVP as in the example above.
+
+![img.png](img/exec_complex_vp.png)
+
+**Step 1-2**: Alice and Bob create their intentVPs and initial partial transactions. Alice spends both of the notes she is ready to give away, expecting to receive one of them back (the other one will go to the counterparty).
+
+**Step 3**: A solver sees Alice's and Bob's partial transactions and matches them together. Alice receives the blue dolphin NFT and one of her notes ([1] of A) back, her intentVP is satisfied and Alice's intent token note is spent. Bob get's [2]B note from Alice, his intentVP is satisfied, his intent token note is spent. All total per-token balances are equal to 0, and the final transaction can be created.
+
+**Step 4**: The final transaction is created using the spent and output notes from the partial transactions.
+
+### 3. One way to represent arbitrary states
+
+For an arbitrary application, notes store the state of the application. When the application needs to change its state, it (i.e. any party that has the authority) can spend the old state and produce a new state.
+If the state change is possible within one partial transaction, such a note has a value 0 and doesn't affect the total balance.
```

Comment:
I'm not sure that I understand the distinction here - is state _not_ consumed in this case? Wouldn't it still be invalid to consume the old note twice?

Reply:
The old state is consumed (once) in all cases. Zero-value state notes in this case are not dummy, so if the nullifier is published, the note can't be spent twice, as I understand

In both cases (state changed within one transaction or state changed across multiple partial transactions) the old state and the new state notes have zero value, but in the latter case intermediate state notes have non-zero value (to make sure the tx is unbalanced until the final state is computed)

Reply:
ah I see, makes sense